### PR TITLE
Use SmallVec to optimize for small integer sizes.

### PR DESCRIPTION
**Pull request by Lemmih**
This is a proof of concept. I'd like to discuss possible designs and pathways to getting something like this merged.

# Motivation

Using `BigInt` with mostly small (<= word-sized) integers is prohibitively expensive. One numerically intense algorithm from rgeometry is 1000 times slower when using `BigRational` instead of a fixed-precision number.

# Performance

`num-bigint` performs quite well for large input and is competitive with the GMP. Detailed benchmarks can be found here: https://github.com/Lemmih/num-criterion

So why do small integers perform so poorly? Mostly because allocating new memory is significantly more expensive than doing a single addition or multiplication. There are also a few cases where we can use specialized algorithms. For example, there's a particularly fast `gcd` algorithm that only works on word-sized integers.

# SmallVec with two inlined BigDigits

`BigUint` is a `Vec<BigDigit>`. As such, it has a size of 3 words in addition to the actual digits. Switching to a `SmallVec<[BigDigit; 2]>` will not increase the size of a `BigUint` but allows for two BigDigits to be inlined and thus not require a heap allocation.

# Summary of benchmarks:

I've run benchmarks on an M1 Apple MacBook Air and a 3950X AMD desktop. The results I get from these platforms are quite different and I'd love it if other people could help out by running the benchmark on their machines. For instructions, see: https://github.com/Lemmih/num-criterion

 - **GCD**: Slightly faster across all bit sizes. This is mainly because I wrote a more efficient version of `shr_assign`. The previous version interacted poorly with `SmallVec`.
 - **Mul**: Significantly faster for bit sizes <=64. Slightly slower for 128-bit integers. No significant difference for larger integers.
 - **Div**: Significantly faster for bit sizes <=128. No significant difference for larger integers.
 - **Add**: Significantly faster for bit sizes <=128. No significant difference for larger integers.
 - **Clone**: Significantly faster for bit sizes <=128. No significant difference for larger integers.

# Other optimizations

## Specialize for integers that fit in 128 bits.

Ideally, addition, multiplication, and division for integers that fit in 128 bits would be as fast as cloning. Copying the bytes is significantly slower than doing the arithmetic.

 * Signed integer addition/subtraction is quite naive. It is comparing the absolute numbers before subtracting/adding. This has to be improved before specialization will have any noticeable effect. Surely we can get within 2x of GMP.

## Use better GCD implementation.

The GCD implementation in `num-integer` is fairly slow and can be significantly improved by changing a line or two. On my system, this leads to a 10x speed improvement for BigInts when the integer fits in 64 bits.

# Other oddities

* Rug and Ramp are really good at cloning data. How can they clone large integers at more than twice the speed of `num-bigint`? Aren't we all just calling 'memcpy'?

### CODE REVIEW

**Comment by Lemmih** [comment]
Here are the raw benchmark results. Red lines are for `Vec`, green lines are for `SmallVec`. All in all, no major surprises expect for `factorial` being 10x faster, `fib2` being slightly slower, and `pow` being a few times faster.

```diff
< test multiply/0 ... bench:          52 ns/iter (+/- 0)
< test multiply/1 ... bench:        4267 ns/iter (+/- 40)
< test multiply/2 ... bench:      335554 ns/iter (+/- 1598)
< test multiply/3 ... bench:      741188 ns/iter (+/- 4529)
> test multiply/0 ... bench:          57 ns/iter (+/- 0)
> test multiply/1 ... bench:        4282 ns/iter (+/- 43)
> test multiply/2 ... bench:      295470 ns/iter (+/- 2401)
> test multiply/3 ... bench:      684280 ns/iter (+/- 4178)

< test divide/0 ... bench:         155 ns/iter (+/- 3)
< test divide/1 ... bench:        2082 ns/iter (+/- 5)
< test divide/2 ... bench:      109332 ns/iter (+/- 535)
< test divide/big_little ... bench:        8534 ns/iter (+/- 243)
> test divide/0 ... bench:          93 ns/iter (+/- 0)
> test divide/1 ... bench:        2157 ns/iter (+/- 27)
> test divide/2 ... bench:      112321 ns/iter (+/- 278)
> test divide/big_little ... bench:        8446 ns/iter (+/- 33)

< test remainder/0 ... bench:         150 ns/iter (+/- 3)
< test remainder/1 ... bench:        2094 ns/iter (+/- 33)
< test remainder/2 ... bench:      110769 ns/iter (+/- 2146)
< test remainder/big_little ... bench:        8573 ns/iter (+/- 363)
> test remainder/0 ... bench:          93 ns/iter (+/- 0)
> test remainder/1 ... bench:        2215 ns/iter (+/- 27)
> test remainder/2 ... bench:      110185 ns/iter (+/- 1028)
> test remainder/big_little ... bench:        8437 ns/iter (+/- 36)

< test factorial_100 ... bench:       10390 ns/iter (+/- 107)
> test factorial_100 ... bench:         995 ns/iter (+/- 3)

< test fib/100 ... bench:         491 ns/iter (+/- 4)
< test fib/1000 ... bench:        8321 ns/iter (+/- 24)
< test fib/10000 ... bench:      666871 ns/iter (+/- 2149)
> test fib/100 ... bench:         554 ns/iter (+/- 3)
> test fib/1000 ... bench:        8703 ns/iter (+/- 29)
> test fib/10000 ... bench:      677992 ns/iter (+/- 5881)

< test fib2/100 ... bench:         690 ns/iter (+/- 9)
< test fib2/1000 ... bench:       16830 ns/iter (+/- 53)
< test fib2/10000 ... bench:     1254097 ns/iter (+/- 12160)
> test fib2/100 ... bench:        1032 ns/iter (+/- 6)
> test fib2/1000 ... bench:       16998 ns/iter (+/- 64)
> test fib2/10000 ... bench:     1245658 ns/iter (+/- 18224)

< test fac_to_string ... bench:         749 ns/iter (+/- 2)
> test fac_to_string ... bench:         857 ns/iter (+/- 7)

< test fib_to_string ... bench:         189 ns/iter (+/- 2)
> test fib_to_string ... bench:         156 ns/iter (+/- 5)

< test to_str_radix/02 ... bench:        2445 ns/iter (+/- 26)
< test to_str_radix/08 ... bench:         909 ns/iter (+/- 2)
< test to_str_radix/10 ... bench:        2105 ns/iter (+/- 3)
< test to_str_radix/16 ... bench:         675 ns/iter (+/- 7)
< test to_str_radix/36 ... bench:        1753 ns/iter (+/- 3)
> test to_str_radix/02 ... bench:        2403 ns/iter (+/- 6)
> test to_str_radix/08 ... bench:         907 ns/iter (+/- 3)
> test to_str_radix/10 ... bench:        2257 ns/iter (+/- 6)
> test to_str_radix/16 ... bench:         696 ns/iter (+/- 1)
> test to_str_radix/36 ... bench:        1930 ns/iter (+/- 17)

< test from_str_radix/from_str_radix_02 ... bench:        2718 ns/iter (+/- 24)
< test from_str_radix/from_str_radix_08 ... bench:        1046 ns/iter (+/- 10)
< test from_str_radix/from_str_radix_10 ... bench:         964 ns/iter (+/- 14)
< test from_str_radix/from_str_radix_16 ... bench:         765 ns/iter (+/- 1)
< test from_str_radix/from_str_radix_36 ... bench:         747 ns/iter (+/- 2)
> test from_str_radix/from_str_radix_02 ... bench:        2661 ns/iter (+/- 10)
> test from_str_radix/from_str_radix_08 ... bench:        1030 ns/iter (+/- 2)
> test from_str_radix/from_str_radix_10 ... bench:         972 ns/iter (+/- 19)
> test from_str_radix/from_str_radix_16 ... bench:         790 ns/iter (+/- 12)
> test from_str_radix/from_str_radix_36 ... bench:         746 ns/iter (+/- 2)

< test rand/64 ... bench:          32 ns/iter (+/- 0)
< test rand/256 ... bench:          39 ns/iter (+/- 0)
< test rand/1009 ... bench:          74 ns/iter (+/- 0)
< test rand/2048 ... bench:         167 ns/iter (+/- 1)
< test rand/4096 ... bench:         272 ns/iter (+/- 7)
< test rand/8192 ... bench:         487 ns/iter (+/- 3)
< test rand/65536 ... bench:        3851 ns/iter (+/- 13)
< test rand/131072 ... bench:        7516 ns/iter (+/- 22)
> test rand/64 ... bench:          36 ns/iter (+/- 0)
> test rand/256 ... bench:          41 ns/iter (+/- 0)
> test rand/1009 ... bench:          75 ns/iter (+/- 0)
> test rand/2048 ... bench:         154 ns/iter (+/- 6)
> test rand/4096 ... bench:         274 ns/iter (+/- 1)
> test rand/8192 ... bench:         487 ns/iter (+/- 2)
> test rand/65536 ... bench:        3852 ns/iter (+/- 32)
> test rand/131072 ... bench:        7518 ns/iter (+/- 67)

< test shl ... bench:         533 ns/iter (+/- 10)
> test shl ... bench:         753 ns/iter (+/- 6)

< test shr ... bench:         641 ns/iter (+/- 5)
> test shr ... bench:         597 ns/iter (+/- 2)

< test hahs ... bench:       78634 ns/iter (+/- 753)
> test hahs ... bench:       85428 ns/iter (+/- 4279)

< test pow_bench ... bench:     5108910 ns/iter (+/- 34772)
> test pow_bench ... bench:     1634661 ns/iter (+/- 37123)

< test pow/pow_bench_bigexp ... bench:     5078143 ns/iter (+/- 89062)
< test pow/1e1000 ... bench:        1717 ns/iter (+/- 8)
< test pow/1e10000 ... bench:       77207 ns/iter (+/- 197)
< test pow/1e100000 ... bench:     2781886 ns/iter (+/- 8826)
< test pow/modpow ... bench:     3718827 ns/iter (+/- 11734)
< test pow/modpow_even ... bench:    10142237 ns/iter (+/- 13493)
> test pow/pow_bench_bigexp ... bench:     1676967 ns/iter (+/- 39520)
> test pow/1e1000 ... bench:        1256 ns/iter (+/- 3)
> test pow/1e10000 ... bench:       47898 ns/iter (+/- 97)
> test pow/1e100000 ... bench:     1588090 ns/iter (+/- 24192)
> test pow/modpow ... bench:     3856001 ns/iter (+/- 13228)
> test pow/modpow_even ... bench:    10226758 ns/iter (+/- 28548)

test iters/to_u32_digits ... bench:         108 ns/iter (+/- 0)
< test iters/iter_u32_digits ... bench:          55 ns/iter (+/- 0)
< test iters/to_u64_digits ... bench:          79 ns/iter (+/- 0)
< test iters/iter_u64_digits ... bench:          43 ns/iter (+/- 0)
> test iters/iter_u32_digits ... bench:          51 ns/iter (+/- 0)
> test iters/to_u64_digits ... bench:          68 ns/iter (+/- 4)
> test iters/iter_u64_digits ... bench:          44 ns/iter (+/- 0)

< test factorial/mul_biguint ... bench:      154595 ns/iter (+/- 332)
< test factorial/mul_u32 ... bench:       43612 ns/iter (+/- 66)
< test factorial/div_biguint ... bench:      506817 ns/iter (+/- 1837)
< test factorial/div_u32 ... bench:      454630 ns/iter (+/- 1166)
> test factorial/mul_biguint ... bench:       44531 ns/iter (+/- 117)
> test factorial/mul_u32 ... bench:       43529 ns/iter (+/- 139)
> test factorial/div_biguint ... bench:      458975 ns/iter (+/- 2308)
> test factorial/div_u32 ... bench:      455520 ns/iter (+/- 849)

< test gcd_euclid/0064 ... bench:        1826 ns/iter (+/- 32)
< test gcd_euclid/0256 ... bench:       27013 ns/iter (+/- 119)
< test gcd_euclid/1024 ... bench:      150825 ns/iter (+/- 1289)
< test gcd_euclid/4096 ... bench:      857305 ns/iter (+/- 2175)
> test gcd_euclid/0064 ... bench:         401 ns/iter (+/- 4)
> test gcd_euclid/0256 ... bench:       16033 ns/iter (+/- 72)
> test gcd_euclid/1024 ... bench:      118497 ns/iter (+/- 395)
> test gcd_euclid/4096 ... bench:      779563 ns/iter (+/- 4492)

< test gcd_stein/0064 ... bench:         747 ns/iter (+/- 12)
< test gcd_stein/0256 ... bench:        3156 ns/iter (+/- 42)
< test gcd_stein/1024 ... bench:       15836 ns/iter (+/- 94)
< test gcd_stein/4096 ... bench:      163188 ns/iter (+/- 467)
> test gcd_stein/0064 ... bench:         679 ns/iter (+/- 8)
> test gcd_stein/0256 ... bench:        2757 ns/iter (+/- 17)
> test gcd_stein/1024 ... bench:       14875 ns/iter (+/- 68)
> test gcd_stein/4096 ... bench:      163034 ns/iter (+/- 359)

< test sqrt/64 ... bench:          29 ns/iter (+/- 0)
< test sqrt/1k ... bench:        2186 ns/iter (+/- 5)
< test sqrt/2k ... bench:        5250 ns/iter (+/- 11)
< test sqrt/4k ... bench:       10412 ns/iter (+/- 23)
> test sqrt/64 ... bench:           4 ns/iter (+/- 0)
> test sqrt/1k ... bench:        2335 ns/iter (+/- 8)
> test sqrt/2k ... bench:        5451 ns/iter (+/- 23)
> test sqrt/4k ... bench:       10699 ns/iter (+/- 41)

< test cbrt/64 ... bench:          36 ns/iter (+/- 0)
< test cbrt/1k ... bench:        2936 ns/iter (+/- 13)
< test cbrt/2k ... bench:        5360 ns/iter (+/- 51)
< test cbrt/4k ... bench:       11655 ns/iter (+/- 74)
> test cbrt/64 ... bench:           8 ns/iter (+/- 0)
> test cbrt/1k ... bench:        2874 ns/iter (+/- 17)
> test cbrt/2k ... bench:        5374 ns/iter (+/- 56)
> test cbrt/4k ... bench:       11488 ns/iter (+/- 77)

< test cbrt/64_nth_10 ... bench:          53 ns/iter (+/- 0)
< test cbrt/1k_nth_10 ... bench:        2738 ns/iter (+/- 52)
< test cbrt/1k_nth_100 ... bench:        1166 ns/iter (+/- 2)
< test cbrt/1k_nth_1000 ... bench:        5212 ns/iter (+/- 10)
< test cbrt/1k_nth_10000 ... bench:          25 ns/iter (+/- 0)
> test cbrt/64_nth_10 ... bench:          17 ns/iter (+/- 0)
> test cbrt/1k_nth_10 ... bench:        2301 ns/iter (+/- 25)
> test cbrt/1k_nth_100 ... bench:         826 ns/iter (+/- 2)
> test cbrt/1k_nth_1000 ... bench:        3160 ns/iter (+/- 9)
> test cbrt/1k_nth_10000 ... bench:           3 ns/iter (+/- 0)

< test cbrt/2k_nth_10 ... bench:        5226 ns/iter (+/- 18)
< test cbrt/2k_nth_100 ... bench:        5721 ns/iter (+/- 96)
< test cbrt/2k_nth_1000 ... bench:        7743 ns/iter (+/- 44)
< test cbrt/2k_nth_10000 ... bench:          25 ns/iter (+/- 0)
> test cbrt/2k_nth_10 ... bench:        4958 ns/iter (+/- 51)
> test cbrt/2k_nth_100 ... bench:        4273 ns/iter (+/- 9)
> test cbrt/2k_nth_1000 ... bench:        5389 ns/iter (+/- 19)
> test cbrt/2k_nth_10000 ... bench:           3 ns/iter (+/- 0)

< test cbrt/4k_nth_10 ... bench:       12697 ns/iter (+/- 49)
< test cbrt/4k_nth_100 ... bench:       11007 ns/iter (+/- 31)
< test cbrt/4k_nth_1000 ... bench:       46102 ns/iter (+/- 213)
< test cbrt/4k_nth_10000 ... bench:          25 ns/iter (+/- 0)
> test cbrt/4k_nth_10 ... bench:       12151 ns/iter (+/- 106)
> test cbrt/4k_nth_100 ... bench:        9124 ns/iter (+/- 23)
> test cbrt/4k_nth_1000 ... bench:       36809 ns/iter (+/- 320)
> test cbrt/4k_nth_10000 ... bench:           3 ns/iter (+/- 0)

test gcd/64bits/uint ... bench:          89 ns/iter (+/- 0)
< test gcd/64bits/branchless ... bench:          78 ns/iter (+/- 1)
< test gcd/64bits/num ... bench:         782 ns/iter (+/- 3)
> test gcd/64bits/branchless ... bench:          78 ns/iter (+/- 0)
> test gcd/64bits/num ... bench:         651 ns/iter (+/- 1)

< test gcd/128/uint ... bench:         557 ns/iter (+/- 2)
< test gcd/128/branchless ... bench:         510 ns/iter (+/- 2)
< test gcd/128/num ... bench:        1591 ns/iter (+/- 8)
> test gcd/128/uint ... bench:         558 ns/iter (+/- 2)
> test gcd/128/branchless ... bench:         510 ns/iter (+/- 1)
> test gcd/128/num ... bench:        1378 ns/iter (+/- 5)

< test gcd/1024/num ... bench:       17710 ns/iter (+/- 73)
> test gcd/1024/num ... bench:       16625 ns/iter (+/- 79)

< test gcd/4096/num ... bench:      165661 ns/iter (+/- 696)
> test gcd/4096/num ... bench:      165488 ns/iter (+/- 872)

< test gcd/32768/num ... bench:     9507381 ns/iter (+/- 12209)
> test gcd/32768/num ... bench:     9614454 ns/iter (+/- 16885)

test mul/64bits/uint ... bench:           0 ns/iter (+/- 0)
< test mul/64bits/num ... bench:          38 ns/iter (+/- 0)
> test mul/64bits/num ... bench:           5 ns/iter (+/- 0)

test mul/128/uint ... bench:           5 ns/iter (+/- 0)
< test mul/128/num ... bench:          42 ns/iter (+/- 0)
> test mul/128/num ... bench:          36 ns/iter (+/- 1)

< test mul/1024/num ... bench:         254 ns/iter (+/- 7)
> test mul/1024/num ... bench:         249 ns/iter (+/- 5)

< test mul/4096/num ... bench:        3332 ns/iter (+/- 16)
> test mul/4096/num ... bench:        3342 ns/iter (+/- 6)

< test mul/32768/num ... bench:      113352 ns/iter (+/- 434)
> test mul/32768/num ... bench:      106407 ns/iter (+/- 368)

< test mut/64bits/num ... bench:          65 ns/iter (+/- 0)
> test mut/64bits/num ... bench:          11 ns/iter (+/- 0)

< test mut/128/num ... bench:          70 ns/iter (+/- 0)
> test mut/128/num ... bench:          41 ns/iter (+/- 0)

< test mut/1024/num ... bench:         307 ns/iter (+/- 16)
> test mut/1024/num ... bench:         303 ns/iter (+/- 16)

< test mut/4096/num ... bench:        3376 ns/iter (+/- 9)
> test mut/4096/num ... bench:        3389 ns/iter (+/- 8)

< test mut/32768/num ... bench:      111984 ns/iter (+/- 253)
> test mut/32768/num ... bench:      106421 ns/iter (+/- 763)

test div/64bits/uint ... bench:           0 ns/iter (+/- 0)
< test div/64bits/num ... bench:          72 ns/iter (+/- 0)
> test div/64bits/num ... bench:          28 ns/iter (+/- 0)

test div/128/uint ... bench:           5 ns/iter (+/- 0)
< test div/128/num ... bench:         109 ns/iter (+/- 0)
> test div/128/num ... bench:          59 ns/iter (+/- 0)

< test div/1024/num ... bench:         130 ns/iter (+/- 1)
> test div/1024/num ... bench:         126 ns/iter (+/- 1)

< test div/4096/num ... bench:         224 ns/iter (+/- 5)
> test div/4096/num ... bench:         226 ns/iter (+/- 4)

< test div/32768/num ... bench:         934 ns/iter (+/- 13)
> test div/32768/num ... bench:         937 ns/iter (+/- 12)

test add/64bits/uint ... bench:           0 ns/iter (+/- 0)
< test add/64bits/num ... bench:          47 ns/iter (+/- 1)
> test add/64bits/num ... bench:          21 ns/iter (+/- 0)

test add/128/uint ... bench:           5 ns/iter (+/- 0)
< test add/128/num ... bench:          47 ns/iter (+/- 1)
> test add/128/num ... bench:          28 ns/iter (+/- 0)

< test add/1024/num ... bench:          71 ns/iter (+/- 3)
> test add/1024/num ... bench:          72 ns/iter (+/- 2)

< test add/4096/num ... bench:         161 ns/iter (+/- 7)
> test add/4096/num ... bench:         161 ns/iter (+/- 8)

< test add/32768/num ... bench:        1084 ns/iter (+/- 131)
> test add/32768/num ... bench:        1068 ns/iter (+/- 127)

test clone/64bits/uint ... bench:           0 ns/iter (+/- 0)
< test clone/64bits/num ... bench:          23 ns/iter (+/- 0)
> test clone/64bits/num ... bench:           4 ns/iter (+/- 0)

test clone/128/uint ... bench:           5 ns/iter (+/- 0)
< test clone/128/num ... bench:          22 ns/iter (+/- 0)
> test clone/128/num ... bench:           5 ns/iter (+/- 0)

< test clone/1024/num ... bench:          32 ns/iter (+/- 1)
> test clone/1024/num ... bench:          33 ns/iter (+/- 1)

< test clone/4096/num ... bench:          64 ns/iter (+/- 9)
> test clone/4096/num ... bench:          64 ns/iter (+/- 10)

< test clone/32768/num ... bench:         318 ns/iter (+/- 110)
> test clone/32768/num ... bench:         293 ns/iter (+/- 96)
```

**Comment by Lemmih** [comment]
Just realised I ran the benchmarks without the `smallvec/union` feature. Will update the results.

**Comment by Lemmih** [comment]
@cuviper Any thoughts about this PR?

**Comment by cuviper** [comment]
Oh, I glanced at this when it was much bigger, with your `BigIntSmall` etc., and I felt overwhelmed to review it. I see you've now reverted most of that and it looks much more manageable. Could you go ahead and squash your commits down to a smaller series for review?

File: Cargo.toml [hunk 1]
```diff
@@ -1,22 +1,23 @@
 [package]
-authors = ["The Rust Project Developers"]
-description = "Big integer implementation for Rust"
+authors       = ["The Rust Project Developers"]
```
Comment (cuviper) [inline Cargo.toml:2]: Please don't change the formatting incidentally -- stick to your change only.

File: src/bigint.rs [hunk 2]
```diff
@@ -538,7 +540,7 @@ impl IntDigits for BigInt {
         self.data.digits()
     }
     #[inline]
-    fn digits_mut(&mut self) -> &mut Vec<BigDigit> {
+    fn digits_mut(&mut self) -> &mut SmallVec<[BigDigit; BigUint::INLINED]> {
```
Comment (cuviper) [inline src/bigint.rs:543]: I suggest adding an aliased `type BigVec = SmallVec<...>` in one place, so the particular type details aren't duplicated everywhere. The exact name is not important to me, if you'd prefer `BigData` or something else, as long as it's clear.

**Comment by Lemmih** [comment]
> Could you go ahead and squash your commits down to a smaller series for review?

**Comment by Lemmih** [comment]
GitHub didn't like that I forced pushed the squashed commits. I've opened a new PR.

**Pull request closed without merging**